### PR TITLE
Implement the 'single' method to compute SNR

### LINKED ISSUE #3476: Implement SNR_single
context: https://github.com/sct-pipeline/gm-challenge/issues/58

i'll fill up more context when i have time

### PR DESCRIPTION
**Pull request by jcohenadad**
#### GitHub

- [x] I've given this PR a concise, self-descriptive, and meaningful title
- [x] I've linked relevant issues in the PR body
- [x] I've applied [the relevant labels](https://www.neuro.polymtl.ca/software/contributing#pr_labels) to this PR
- [x] I've applied a [release milestone](https://github.com/spinalcordtoolbox/spinalcordtoolbox/milestones) (major, minor, patch) in line with [Semantic Versioning guidelines](https://github.com/spinalcordtoolbox/spinalcordtoolbox/wiki/Misc%3A-Creating-a-new-release#convention-for-naming-releases) 
- [x] I've assigned a reviewer

#### PR contents

- [x] I've consulted [SCT's internal developer documentation](https://github.com/spinalcordtoolbox/spinalcordtoolbox/wiki) to ensure my contribution is in line with any relevant design decisions
- [x] I've added [relevant tests](https://github.com/spinalcordtoolbox/spinalcordtoolbox/wiki/Programming%3A-Tests) for my contribution
- [x] I've updated the [relevant documentation](https://github.com/spinalcordtoolbox/spinalcordtoolbox/wiki/Programming%3A-Documentation) for my changes, including argparse descriptions, docstrings, and ReadTheDocs tutorial pages

## Description

The main purpose of this PR is to implement the 'single' method to compute SNR. More details about the content of the PR below: 

- Implemented the 'single' method to compute SNR, which is based on a single 3D volume. This method compute the noise standard deviation based on a small box at the corner of an image. Fixes #3476 
- Implemented thorough unit tests for all SNR methods.
- Implemented a slice-wise computation of the SNR, to account for inhomogeneous transmit across slices (assuming last dimension) and/or bias field intensity correction filter.
- Clarify the correction factor for the SNR_diff method. Fixes https://github.com/spinalcordtoolbox/spinalcordtoolbox/issues/3481

## Follow-up issues

I am concerned about the bias of the SNR_single estimation: about 25% less than the SNR estimated with the mult and diff methods. However, I double-checked multiple times the equation A12 from [Dietrich et al.](https://paperpile.com/app/p/d1f035d3-0f3c-00b4-9c5a-c4bf326a57ab) and the implementation seems legit. More details in https://github.com/spinalcordtoolbox/spinalcordtoolbox/issues/3486

### CODE REVIEW

**Review by joshuacwnewton — CHANGES_REQUESTED** [review]
Thanks for taking the time to write such a thorough set of tests for the various uses of `sct_compute_snr`! :smile:

File: spinalcordtoolbox/scripts/sct_compute_snr.py
Comment (jcohenadad) [reply spinalcordtoolbox/scripts/sct_compute_snr.py]: great point-- i made it use 0,1 for more convenience in 6e38ed9fb68977a9cfba4f8ea67f8ead99654825
and added a test in e0225ded451f5092fbcade655d95f4fc22fd0120

File: spinalcordtoolbox/scripts/sct_compute_snr.py
Comment (jcohenadad) [reply spinalcordtoolbox/scripts/sct_compute_snr.py]: fixed in 9b58de90bd00a6b261948d4474b967fa5f8a5376

File: spinalcordtoolbox/scripts/sct_compute_snr.py
Comment (jcohenadad) [reply spinalcordtoolbox/scripts/sct_compute_snr.py:139]: Indeed, since we no longer compute SNR slicewise

File: spinalcordtoolbox/scripts/sct_compute_snr.py
Comment (jcohenadad) [reply spinalcordtoolbox/scripts/sct_compute_snr.py:223]: I'm glad you bring that up. I was doing that at the beginning, but I didn't find an elegant way to unit test this case, given that parser.error does not raise anything (or at least I was not able to catch anything with the convenient `with pytest.raises(<ERROR TYPE>):` syntax.

File: testing/cli/test_cli_sct_compute_snr.py
Comment (jcohenadad) [reply testing/cli/test_cli_sct_compute_snr.py]: > but check the message as well. 

yeah... i'm afraid it would add unnecessary time on the core dev team in case someone slightly adjusts the message, then the test would not pass, the person would need to figure out why, etc.

File: testing/cli/test_cli_sct_compute_snr.py
Comment (jcohenadad) [reply testing/cli/test_cli_sct_compute_snr.py:15]: great point-- in 6fc5ffe70be55e7aa3f3b9e1bd3fd330a1f0faee

File: spinalcordtoolbox/scripts/sct_compute_snr.py
Comment (joshuacwnewton) [reply spinalcordtoolbox/scripts/sct_compute_snr.py:223]: > but I didn't find an elegant way to unit test this case, given that parser.error does not raise anything (or at least I was not able to catch anything with the convenient with pytest.raises(<ERROR TYPE>): syntax.

Interesting... that's definitely a significant limitation! Thanks for letting me know.  I'll try and dig a little deeper -- I feel like this is something we should be able to test in theory. :thinking:

**Comment by jcohenadad** [comment]
oopsi! i ran into this error if `-m-noise` is not binary:
~~~
sct_compute_snr -i data_concat.nii.gz -method single -m sub-9710_run-1_T2starw_crop_wmseg_erode.nii.gz -m-noise sub-9710_run-1_T2starw_crop_wmseg_erode.nii.gz -o snr_single.txt
--

Traceback (most recent call last):
  File "/Users/julien/code/sct/spinalcordtoolbox/scripts/sct_compute_snr.py", line 247, in <module>
    main(sys.argv[1:])
  File "/Users/julien/code/sct/spinalcordtoolbox/scripts/sct_compute_snr.py", line 230, in main
    std_in_roi = np.std(data3d[mask_noise])
IndexError: arrays used as indices must be of integer (or boolean) type
~~~

I'll fix it

File: testing/cli/test_cli_sct_compute_snr.py
Comment (joshuacwnewton) [reply testing/cli/test_cli_sct_compute_snr.py]: That's a good point... 

Maybe we wouldn't put the entire message, but a small part of the message that uniquely identifies that error. That way, minor changes can be made with no failure. For example, some of the messages mention a specific argument (e.g. `-m`) so `pytest.raises` could match against just that small string. 

I have a hunch that it might not be worth the effort, though. :thinking:

**Comment by jcohenadad** [comment]
We also need a way to remove the Rayleigh noise correction in case the noise mask is in a high SNR regime (eg: on tissue vs. air)

File: testing/cli/test_cli_sct_compute_snr.py
Comment (jcohenadad) [reply testing/cli/test_cli_sct_compute_snr.py]: sorry i have already spend too much time on this PR (other priorities are burning me)-- so i won't be able to follow up

File: spinalcordtoolbox/scripts/sct_compute_snr.py
Comment (joshuacwnewton) [reply spinalcordtoolbox/scripts/sct_compute_snr.py]: Just to make sure I'm understanding correctly, does `mask_noise` have to be the exact same shape as `data3d`?

6293715 only checks that the mask is 3D, but maybe we could check if `mask_noise.shape == data3d.shape`.

File: testing/cli/test_cli_sct_compute_snr.py
Comment (joshuacwnewton) [reply testing/cli/test_cli_sct_compute_snr.py]: Of course. This is nitpicky anyway, so I will mark as resolved.

File: spinalcordtoolbox/scripts/sct_compute_snr.py
Comment (joshuacwnewton) [reply spinalcordtoolbox/scripts/sct_compute_snr.py]: Ah, I've seen your other comment about not having enough time to address. No worries about answering -- I can take care of the remaining feedback.

File: spinalcordtoolbox/scripts/sct_compute_snr.py
Comment (joshuacwnewton) [reply spinalcordtoolbox/scripts/sct_compute_snr.py:235]: > (or at least I was not able to catch anything with the convenient `with pytest.raises(<ERROR TYPE>):` syntax.

Here is a way to test whether `parser.error` has been thrown: 

```
    with pytest.raises(SystemExit) as e:
        sct_compute_snr.main(argv=['-i', dummy_4d_nib, '-m', dummy_3d_nib, '-method', 'single'])
    assert e.value.code == 2
```

If we wanted, we could also modify our parser to throw a more specific error than `sys.exit(2)`:

https://github.com/spinalcordtoolbox/spinalcordtoolbox/blob/40781c1ac8bdd62f86d2103ea98a6a85f5d8aad4/spinalcordtoolbox/utils/shell.py#L131-L140

But, I think testing against `sys.exit(2)` is good enough? I can go ahead modify all of the arg-related exceptions in this PR to use this pattern.

File: spinalcordtoolbox/scripts/sct_compute_snr.py
Comment (jcohenadad) [reply spinalcordtoolbox/scripts/sct_compute_snr.py]: > Just to make sure I'm understanding correctly, does mask_noise have to be the exact same shape as data3d?

no

File: spinalcordtoolbox/scripts/sct_compute_snr.py
Comment (joshuacwnewton) [inline spinalcordtoolbox/scripts/sct_compute_snr.py:90]: I notice that `-rayleigh` is on by default for all methods (even `'mult'`). Could this be detrimental to users who don't check this carefully?

In other words, would it be better for this to be "opt-in" rather than "opt-out"?

Comment (jcohenadad) [reply spinalcordtoolbox/scripts/sct_compute_snr.py:90]: > In other words, would it be better for this to be "opt-in" rather than "opt-out"?

Rayleigh _should_ be on for diff, and for single (in most cases). And it is not used for mult

Comment (joshuacwnewton) [reply spinalcordtoolbox/scripts/sct_compute_snr.py:90]: Ah, thank you for the clarification. In what cases would someone want to set `-rayleigh 0`, then?

Maybe rather than "`to use it with the 'single' method in case the noise mask is taken in a region with low SNR (e.g., the air).`" we could give an example of when to turn this option off instead (since it's already on by default).

Something like "`This should only be turned off in cases where [...]`".

**Review by joshuacwnewton — APPROVED** [review]
Approving in advance since I believe the only real feedback left is documentation.

File: spinalcordtoolbox/scripts/sct_compute_snr.py
Comment (jcohenadad) [reply spinalcordtoolbox/scripts/sct_compute_snr.py:90]: yes good point

File: spinalcordtoolbox/scripts/sct_compute_snr.py
Comment (jcohenadad) [reply spinalcordtoolbox/scripts/sct_compute_snr.py:139]: @joshuacwnewton actually, I noticed that some volumes have a scaled intensity across slices, meaning that we might want to compute SNR per slice and then average. I'll do more thorough investigations and add specific unit test

**Review by joshuacwnewton — COMMENTED** [review]
Here is my attempt to explain the differences in notation between papers, which resulted in the `sqrt(2)` vs. `1/sqrt(2)` mix-up.

This might not be strictly necessary, but I wanted to confirm my understanding by writing it out.

File: spinalcordtoolbox/scripts/sct_compute_snr.py
Comment (jcohenadad) [reply spinalcordtoolbox/scripts/sct_compute_snr.py:223]: @joshuacwnewton would you prefer I add your explanation? I thought redirecting to this issue would suffice:

> # More context in: https://github.com/spinalcordtoolbox/spinalcordtoolbox/issues/3481

Comment (jcohenadad) [reply spinalcordtoolbox/scripts/sct_compute_snr.py:223]: > This might not be strictly necessary, but I wanted to confirm my understanding by writing things out. 😅

ah! i'm just seeing this. Indeed, your understanding is 100% correct 😊

File: spinalcordtoolbox/scripts/sct_compute_snr.py
Comment (joshuacwnewton) [reply spinalcordtoolbox/scripts/sct_compute_snr.py:95]: Resolved by 8e10b2d.

File: spinalcordtoolbox/scripts/sct_compute_snr.py
Comment (joshuacwnewton) [reply spinalcordtoolbox/scripts/sct_compute_snr.py:223]: Basically just wanted to leave some documentation on this PR for future reference, hehe. :smile: 

I've marked these comments as resolved, and so I think this PR is good to go!

**Pull request merged**
